Xenstore refactoring

LKomaryanskiy opened the pull request:
The PR fixes locks, issues with different path lengths, and other fixes.

firscity commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -922,10 +922,15 @@ static void handle_transaction_stop(struct xen_domain *domain, uint32_t id,
+		return;
+	}
+
+	snprintf(path, XENSTORE_MAX_LOCALPATH_LEN,
```
I'm not sure that this is right way for copying payload. As I understood, `payload` may be not NULL-terminated, but will be size of `len`. Previous implementation handled such case, need to recheck if it needed.

LKomaryanskiy replied:
Yes, in this case it can be an issue. I recheck this and found the solution: we can specify payload len as a string len:
`snprintf(path, XENSTORE_MAX_LOCALPATH_LEN, "/local/domain/%.*s", len, payload);`
This code works fine in all corner cases that I've checked:
1) If payload is less than fittable size and doesn't have null-terminated symbol (e.g. char payload[1] = "1";
2) if payload is bigger than fittable and doesn't have null-terminated symbol (e.g. char payload[20] = "12345678901234567890"
3-4) Both above cases with null-terminated symbol.

So, I'll fix it in this way.

firscity commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -280,33 +282,58 @@ static void send_reply_read(struct xen_domain *domain, uint32_t id,
-	struct xs_entry *entry = key_to_entry(path);
-
+	entry = key_to_entry(path);
 	if (entry) {
```
it is better to make reverse condition here ( `if (!entry) { goto ...}`

LKomaryanskiy replied:
Good point. I'll change this.

firscity commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -558,6 +518,47 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
+int xss_write(const char *path, const char *value)
+{
+	xss_do_write(path, value);
+	notify_watchers(path, DOMID_SELF);
```
DOMID_SELF is not working for this situation. It is a Xen related macro, that makes Xen to get domid of caller domain. Dom0 domid should be used here (just `0`).

Also, please note, that  `_handle_write()` also calls `notify_watchers()` and this function, it should be done only once per write.

firscity commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -558,6 +518,47 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
+{
+	int rc;
+	/* TODO: check this */
+	char ns[32] = { 0 };
```
Substitute with previously added macro for int size.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -357,32 +357,45 @@ static int fire_watcher(struct xen_domain *domain, char *pending_path)
 	char *path;
 	char *tok, *tok_state;
-	size_t vall = strlen(data) + 1;
+	size_t val = strlen(data) + 1;
```
Ok. I'll change this.

LKomaryanskiy commented:
@firscity thank you for your close review. I've updated my PR.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -357,32 +357,45 @@ static int fire_watcher(struct xen_domain *domain, char *pending_path)
-	strcpy(path, const_path);
+	if (!path) {
+		LOG_ERR("Failed to allocate memory for path\n");
+		return;
```
You need to return an error and propagate to it the calling domain.

lorc replied:
Yes, exactly this

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -22,6 +22,8 @@
 LOG_MODULE_REGISTER(xenstore);
 
+/* Max string length of int32_t + terminating null symbol */
+#define INT32_MAX_STR_LEN (11)
```
I'm sorry, but max length if int32_t is 11 even without terminating NULL: `-2147483647`

LKomaryanskiy replied:
Sure, I'll change it to 12. I forget the minus symbol (11 symbols of "-2147483647" + terminating NULL)

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -280,33 +282,61 @@ static void send_reply_read(struct xen_domain *domain, uint32_t id,
+	size_t reply_sz = 0, dir_name_len;
+	char *path;
+	struct xs_entry *entry, *iter;
+	uint32_t path_len = strlen(payload) + 1;
```
We should use `size_t` there, because this is what is returned by `strlen`. Otherwise, compillers in very strict mode can be unhappy.

LKomaryanskiy replied:
Ok, I'll change this.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -560,6 +521,46 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
+	return 0;
+}
+
+int xss_read(const char *path, char *value, size_t len)
```
Spurious change?

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -560,6 +521,46 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
+	return rc;
+}
+
+int xss_read_integer(const char *path, int *value)
```
Spurious change?

LKomaryanskiy replied:
Answered above.

lorc commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -560,6 +521,46 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
+	return rc;
+}
+
+int xss_set_perm(const char *path, domid_t domid, enum xs_perm perm)
```
Spurious change?

LKomaryanskiy replied:
Answered above.

lorc reviewed (commented):
Also for "xenstore-srv: add protection for wather_list entries": fix typo in "wather_list"

LKomaryanskiy commented:
@lorc thank you for your close review. I've pushed a new version with the proposed fixes.

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -250,34 +307,61 @@ static void send_reply_read(struct xen_domain *domain, uint32_t id,
+	if (!dir_list) {
+		LOG_ERR("Failed to allocate memory for dir list");
+		send_reply(domain, id, XS_ERROR, "ENOMEM");
+		return;
```
(question) where did you free `path`?

LKomaryanskiy replied:
Thank you, I missed this. I'll add it in the next PR version.

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -250,34 +307,61 @@ static void send_reply_read(struct xen_domain *domain, uint32_t id,
 	}
 
+out:
 	k_mutex_unlock(&xsel_mutex);
```
(question) where did you free `path`?

LKomaryanskiy replied:
Fixed in the latest version of the PR.

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -384,57 +484,27 @@ void xss_do_write(const char *const_path, const char *data)
+		iter->value = k_malloc(data_len);
+		if (!iter->value) {
+			LOG_ERR("Failed to allocate memory for xs entry value\n");
+			rc = -ENOMEM;
```
(question) should you remove iter from list and destroy all data allocated for it ?

xakep-amatop replied:
could you please show where we use `value` field after this code executes and where is checked that ptr isn't NULL?

LKomaryanskiy replied:
For example `handle_read()`:
`send_reply_read(domain, id, XS_READ, entry->value ? entry->value : "");`
or `xss_do_rm()`:
`if (entry->value) {`
`		k_free(entry->value);`
`		entry->value = NULL;`
`}`

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -384,57 +484,27 @@ void xss_do_write(const char *const_path, const char *data)
 	}
 
-	if (iter && vall > 0) {
+	if (iter && data_len > 0) {
```
(question) is it possible that we allocate few iter-s inside above loop ?
if yes, what is happened if last allocation for iter is failed and iter is qual to NULL
if no, why we continue iteration?

LKomaryanskiy replied:
> is it possible that we allocate few iter-s inside above loop ?

Yes, it can happen in case if we try to write data to the path, that contains subdirectories that were not created before (e.x. if path is "/local/domain/1/memory/videoram" and before we didn't create "memory" node, the loop will create at first "memory" node and then "videoram").

So, I think it's ok and we do not do something illegal in this case, we just, create parent nodes for the desired path.

xakep-amatop replied:
so, if we have a few iter-s, do we need to execute code under this `if` for all of iter-s?

xakep-amatop replied:
so, in case if you alloc mem for a few iter-s but for the last one allocator will fail, e.g. iter is equal to NULL
you don't update value in a last one, is it right?

xakep-amatop replied:
but we stiill can have a problem, it isn't here, but lets imagine
you alloca a few iter-s and then during allocation a new one you can get a mem chunk a go to error handling without executing code under this if for the last one that you have added to a list

LKomaryanskiy replied:
Yes, it could be in case we add a new node and I think it's ok because we will return the error code in this case, which will give information to the caller that something went wrong. Or do you propose to save the node from which we started allocation of new children and free recursively them? @firscity what do you think about it?

NOTE: I was searching the information on how it should be, but I didn't find such details in xenstore documentation (only this description: `WRITE creates any missing parent paths, with empty values.`) and unfortunately original xenstore code has different organization, so it's not clear how it should be.

LKomaryanskiy replied:
Ok, thank you. I think we will discuss about this place later (fortunately our conversation will be saved in github history)

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -482,35 +552,102 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
+
+	rc = xss_read(path, ns, sizeof(ns));
+	if (!rc)
+		*value = atoi(ns);
```
(note) atoi function may fail and returns zero, at least if we are talking about function from Standard C library

LKomaryanskiy replied:
@xakep-amatop I'm sorry, I'm trying to fix this, but Zephyr doesn't implement sscanf function and as I see in all places used atoi.. So, I think I leave it as is.

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -482,35 +552,102 @@ static void notify_watchers(const char *path, uint32_t caller_domid)
 
-	if (len < data_offset) {
+	data = payload + path_len;
+	if (len < path_len) {
```
(note/not your issue) looks like we have to use here `<=` instead of `<` or change below logic a little bit
lets imagine that `len` is equal to `path_len`, `data` points to byte after `\0` --- > `data = payload + path_len`
and then we write to `data[len - path_len] = 0;` --> `data[0]` byte --> it is equal to `*(payload + payload len + 1) = 0;`
so, looks like we have a mem corrupt here

LKomaryanskiy replied:
I rechecked these lines and I think it should be fixed in some different way because for example, if we pass payload such as ['1', '2', '3', '4', '5', '\0', '9'] (payload without terminating NULL) `path_len` will be 6, `len` - 7 and we will write to data[1] that placed in the out of the given memory. Moreover, it's not a good idea to change the original payload. I think I will fix it in different PR, because these changes can drag review of current PR.

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -597,23 +734,43 @@ static void handle_reset_watches(struct xen_domain *domain, uint32_t id,
-	char path[STRING_LENGTH_MAX];
+	char *path;
 	struct xs_entry *entry;
+	size_t path_len = strlen(payload) + 1;
```
(note) maybe we can make a small static inline function, I see code duplicates in this and above functions

LKomaryanskiy replied:
Yeap, I will add helper inline function for this and additionally, we can add check if payload is not NULL.

xakep-amatop commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -653,56 +817,65 @@ static int xss_do_rm(struct xs_entry *entry)
-	bool path_is_relative;
-
-	path_is_relative = !!(memcmp(payload, localpath, strlen(localpath)));
+	bool path_is_relative = !is_abs_path(payload);
```
(note) just use `is_root_path` function here

LKomaryanskiy replied:
`is_root_path()` check the length of the given path and will return false for example for "/vm/..." which is not correct. So I think using `is_abs_path()` in this case is right.

xakep-amatop replied:
ok, but names of the functions (is_root_path & is_abs_path) is a little but confusing for me...

LKomaryanskiy replied:
Hmm, do you have any idea how to name it better? In my opinion, "is_root" means the path is exactly the root ("/").

xakep-amatop replied:
oooh, I just misunderstood sorry, I thought about relative and absolute path

xakep-amatop reviewed (changes requested):
https://github.com/xen-troops/zephyr-xenlib/pull/33/files#r1172287026

LKomaryanskiy commented:
@xakep-amatop thank you for your close review. I've fixed the memory leak with "path" that you found and added helper inline function in order to calculate sting length with termination NULL symbol. Other issues, I will fix in the further PRs.

LKomaryanskiy commented on xenstore-srv/src/xenstore_srv.c:
```diff
@@ -99,27 +130,63 @@ struct watch_entry *key_to_watcher(char *key, bool complete, char *token)
+/*
+ * Returns the size of string including terminating NULL symbol.
+ */
+static inline size_t str_byte_size(const char *str)
```
Oh, understood. I have an idea of how to reduce boilerplate code. Maybe is it worth making a function that will unpack path from payload, check its size, allocate memory for it and return a pointer? I think it can reduce the amount of code that does the same things.

LKomaryanskiy replied:
Thanks, you pay attention to duplicate code. I remastered it to make the same things reusable.

LKomaryanskiy commented:
In the latest PR version, fixed small review findings, fixed warnings from checkpatch, and removed boilerplate code.

LKomaryanskiy commented:
Rebased to main branch base.

firscity reviewed (approved):
`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`

LKomaryanskiy commented:
Added @firscity  RB

lorc commented:
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

LKomaryanskiy commented:
Added @lorc acked-by

xakep-amatop reviewed (approved):
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`
